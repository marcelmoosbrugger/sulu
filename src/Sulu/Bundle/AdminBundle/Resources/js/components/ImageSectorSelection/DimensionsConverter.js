// @flow
import type {SelectionData} from '../RectangleSelection/types';

export default class DimensionsConverter {
    computedWidth: number;
    computedHeight: number;
    realWidth: number;
    realHeight: number;

    constructor(computedWidth: number, computedHeight: number, realWidth: number, realHeight: number) {
        this.computedWidth = computedWidth;
        this.computedHeight = computedHeight;
        this.realWidth = realWidth;
        this.realHeight = realHeight;
    }

    computedHorizontalToReal(horizontal: number): number {
        return horizontal * this.realWidth / this.computedWidth;
    }

    computedVerticalToReal(vertical: number): number {
        return vertical * this.realHeight / this.computedHeight;
    }

    computedDataToReal(data: SelectionData): SelectionData {
        return {
            width: this.computedHorizontalToReal(data.width),
            height: this.computedVerticalToReal(data.height),
            left: this.computedHorizontalToReal(data.left),
            top: this.computedVerticalToReal(data.top),
        };
    }

    realDataToComputed(data: SelectionData): SelectionData {
        return {
            width: data.width * this.computedWidth / this.realWidth,
            height: data.height * this.computedHeight / this.realHeight,
            left: data.left * this.computedWidth / this.realWidth,
            top: data.top * this.computedWidth / this.realWidth,
        };
    }
}
