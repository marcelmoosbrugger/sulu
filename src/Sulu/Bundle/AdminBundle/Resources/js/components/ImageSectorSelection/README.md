This component builds upon the RectangleSelection and defines more specific functionality
useful when dealing with large images.
Imagine loading a large 1920x1080 image and scaling it down to 640x360 via CSS.
When selecting a sector of the image, it is (depending on the use case) desired to
get the coordinates of the selection (width, height, top and left) with respect to
the original dimensions of the image and not the scaled down ones.
This is exactly what this component does.

```
<div style={{width: 640, height: 360}}>
    <ImageSectorSelection
        initialSelection={{width: 1000, height: 800, top: 200, left: 300}}
        imageSrc="https://unsplash.it/1920/1080" />
</div>
```

Like with the RectangleSelection, if both the minWidth and minHeight properties are set,
the ratio between this to is enforced on the selection.

```
initialState = {selection: {}};
<div>
    <div style={{width: 640, height: 360}}>
        <ImageSectorSelection
            initialSelection={{width: 1500, height: 800, top: 200, left: 300}}
            minWidth={100}
            minHeight={60}
            imageSrc="https://unsplash.it/1920/1080"
            onChange={s => setState({selection: s})}
        />
    </div>
    
    <p>
        Width: {state.selection.width}, 
        Height: {state.selection.height}, 
        Top: {state.selection.top}, 
        Left: {state.selection.left}
    </p>
</div>
```
