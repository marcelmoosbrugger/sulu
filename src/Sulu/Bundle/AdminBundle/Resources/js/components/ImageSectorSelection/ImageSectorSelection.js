// @flow
import {action, computed, observable} from 'mobx';
import DimensionsConverter from './DimensionsConverter';
import React from 'react';
import RectangleSelection from '../RectangleSelection';
import RoundingNormalizer from '../RectangleSelection/dataNormalizers/RoundingNormalizer';
import type {SelectionData} from '../RectangleSelection/types';
import log from 'loglevel';
import {observer} from 'mobx-react';
import selectionStyles from './imageSectorSelection.scss';

@observer
export default class ImageSectorSelection extends React.PureComponent {
    props: {
        /** Determines the position at which the selection box is rendered at the beginning. */
        initialSelection?: SelectionData,
        minWidth?: number,
        minHeight?: number,
        onChange?: (s: SelectionData) => void,
        imageSrc: string,
    };

    image: Image;
    rounding = new RoundingNormalizer();

    @observable imageLoaded = false;
    @observable containerWidth: number;
    @observable containerHeight: number;

    @action componentWillMount() {
        this.image = new Image();
        this.image.onload = action(() => this.imageLoaded = true);
        this.image.onerror = () => log.error('Failed to preload image');
        this.image.src = this.props.imageSrc;
    }

    @computed get converter(): DimensionsConverter {
        return new DimensionsConverter(
            this.image.naturalWidth, this.image.naturalHeight, this.imageResizedWidth, this.imageResizedHeight);
    }

    @computed get imageResizedHeight(): number {
        if (this.imageTouchesHoriziontalBorders()) {
            return Math.min(this.image.naturalHeight, this.containerHeight);
        } else {
            return this.imageResizedWidth * this.image.naturalHeight / this.image.naturalWidth;
        }
    }

    @computed get imageResizedWidth(): number {
        if (this.imageTouchesHoriziontalBorders()) {
            return this.imageResizedHeight * this.image.naturalWidth / this.image.naturalHeight;
        } else {
            return Math.min(this.image.naturalWidth, this.containerWidth);
        }
    }

    imageTouchesHoriziontalBorders() {
        let imageHeightToWidth = this.image.naturalHeight / this.image.naturalWidth;
        let containerHeightToWidth = this.containerHeight / this.containerWidth;
        return imageHeightToWidth > containerHeightToWidth;
    }

    handleRectangleSelectionChange = (data: SelectionData) => {
        const computedData = this.rounding.normalize(this.converter.realDataToComputed(data));
        if (this.props.onChange) {
            this.props.onChange(computedData);
        }
    };

    readContainerDimensions = (container: HTMLElement) => {
        if (!container) return;
        window.requestAnimationFrame(action(() => {
            this.containerWidth = container.clientWidth;
            this.containerHeight = container.clientHeight;
        }));
    };

    render() {
        if (!this.imageLoaded || !this.containerWidth || !this.containerHeight) {
            return this.renderContainer();
        } else {
            return this.renderSelection();
        }
    }

    renderContainer(content: ?React.Element<*>) {
        return <div ref={this.readContainerDimensions} className={selectionStyles.selection}>{content}</div>;
    }

    renderSelection() {
        let minWidth, minHeight, initialSelection;
        if (this.props.minWidth) {
            minWidth = this.converter.computedHorizontalToReal(this.props.minWidth);
        }
        if (this.props.minHeight) {
            minHeight = this.converter.computedVerticalToReal(this.props.minHeight);
        }
        if (this.props.initialSelection) {
            initialSelection = this.converter.computedDataToReal(this.props.initialSelection);
        }
        return this.renderContainer(
            <div>
                <RectangleSelection
                    initialSelection={initialSelection}
                    minWidth={minWidth}
                    minHeight={minHeight}
                    onChange={this.handleRectangleSelectionChange}
                    round={false}>
                    <img
                        width={this.imageResizedWidth}
                        height={this.imageResizedHeight}
                        src={this.props.imageSrc} />
                </RectangleSelection>
            </div>
        );
    }
}
