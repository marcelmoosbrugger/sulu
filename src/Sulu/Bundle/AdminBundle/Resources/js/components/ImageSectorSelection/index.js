// @flow
import ImageSectorSelection from './ImageSectorSelection';

export default ImageSectorSelection;
