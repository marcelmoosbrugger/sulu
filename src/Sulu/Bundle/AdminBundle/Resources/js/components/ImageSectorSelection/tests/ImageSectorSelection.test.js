/* eslint-disable flowtype/require-valid-file-annotation */
import {mount, render, shallow} from 'enzyme';
import ImageSectorSelection from '../ImageSectorSelection';
import React from 'react';
import {action} from 'mobx';
import {observer} from 'mobx-react';

@observer
class MockedImageSelection extends ImageSectorSelection {
    constructor(props) {
        super(props);
        this.spyed = false;
    }

    componentDidUpdate() {
        if (!this.spyed && this.props.updateSpy) {
            this.spyed = true;
            // Move the update spy to the end of the execution queue, in order to see rendering changes
            Promise.resolve().then(this.props.updateSpy);
        }
        super.componentDidUpdate();
    }

    componentWillMount() {
        this.image = {naturalWidth: 1920, naturalHeight: 1080};
        this.imageLoaded = true;
    }

    readContainerDimensions = action(() => {
        this.containerWidth = 640;
        this.containerHeight = 360;
    });
}

jest.mock('../../RectangleSelection', () => {
    const RectangleSelection = require.requireActual('../../RectangleSelection').default;
    const {observer} = require.requireActual('mobx-react');
    const {action} = require.requireActual('mobx');

    @observer
    class MockedRectangleSelection extends RectangleSelection {
        readContainerDimensions = action(() => {
            this.container = {clientWidth: 640, clientHeight: 360};
            this.containerHeight = this.container.clientHeight;
            this.containerWidth = this.container.clientWidth;
        });
    }

    return MockedRectangleSelection;
});

test('The component should render with image source', () => {
    const view = render(<ImageSectorSelection imageSrc="//:0" />);
    expect(view).toMatchSnapshot();
});

test('The component should calculate the selection with respect to the image', (done) => {
    window.requestAnimationFrame = jest.fn((cb) => cb());

    const onChangeSpy = (data) => {
        expect(data).toEqual({width: 1920, height: 1080, top: 0, left: 0});
        done();
    };

    mount(
        <MockedImageSelection
            onChange={onChangeSpy}
            imageSrc="//:0" />
    );
});

test('The component should render with initial selection', (done) => {
    window.requestAnimationFrame = jest.fn((cb) => cb());

    const spy = () => {
        expect(view.render()).toMatchSnapshot();
        done();
    };
    const onChangeSpy = (data) => {
        expect(data).toEqual({ width: 1500, height: 800, top: 200, left: 300 });
    };

    const view = mount(
        <MockedImageSelection
            onChange={onChangeSpy}
            updateSpy={spy}
            imageSrc="//:0"
            initialSelection={{width: 1500, height: 800, top: 200, left: 300}} />
    );
});

test('The component should render with minWidth and minHeight', (done) => {
    window.requestAnimationFrame = jest.fn((cb) => cb());

    const spy = () => {
        // don't render the view to see if the minWidth and the minHeight props get passed correctly
        expect(view).toMatchSnapshot();
        done();
    };

    const view = mount(
        <MockedImageSelection
            updateSpy={spy}
            imageSrc="//:0"
            minHeight={300}
            minWidth={600} />
    );
});
