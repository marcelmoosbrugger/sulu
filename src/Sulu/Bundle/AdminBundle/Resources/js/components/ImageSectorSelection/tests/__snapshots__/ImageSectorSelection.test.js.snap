// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render with image source 1`] = `
<div
  class="selection"
/>
`;

exports[`The component should render with initial selection 1`] = `
<div
  class="selection"
>
  <div>
    <div
      class="selection"
    >
      <img
        height="360"
        src="//:0"
        width="640"
      />
      <div
        class="rectangle"
        style="left: 100px; top: 66.66666666666667px; width: 500px; height: 266.6666666666667px;"
      >
        <div
          class="resizeHandle"
        />
        <div
          class="backdrop"
          style="outline-width: 640px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`The component should render with minWidth and minHeight 1`] = `
<MockedImageSelection
  minHeight={300}
  minWidth={600}
  src="//:0"
  updateSpy={[Function]}
>
  <div
    className="selection"
  >
    <div>
      <MockedRectangleSelection
        minHeight={100}
        minWidth={200}
        onChange={[Function]}
        round={false}
      >
        <div
          className="selection"
        >
          <img
            height={360}
            src="//:0"
            width={640}
          />
          <ModifiableRectangle
            backdropSize={640}
            height={320}
            left={0}
            onChange={[Function]}
            onDoubleClick={[Function]}
            top={20}
            width={640}
          >
            <div
              className="rectangle"
              onDoubleClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "height": "320px",
                  "left": "0px",
                  "top": "20px",
                  "width": "640px",
                }
              }
            >
              <div
                className="resizeHandle"
                onMouseDown={[Function]}
              />
              <div
                className="backdrop"
                style={
                  Object {
                    "outlineWidth": "640px",
                  }
                }
              />
            </div>
          </ModifiableRectangle>
        </div>
      </MockedRectangleSelection>
    </div>
  </div>
</MockedImageSelection>
`;
